### ISSUE TITLE
PageListViewportGestures onLongTapStart callback introduces a delay to tiny swipes

### ISSUE THREAD

**Issue by dodgog** [Flutter-Bounty-Hunters/page_list_viewport#54]
When the user executes tiny swipes (imagine a flick of approximately 10 pixels of translation distance) the viewport will seemingly stutter and start the scrolling movement with a noticeable delay of the order of 100ms. See the loom below:

https://www.loom.com/share/82d9eeff4f1c4928825fcfb489c13a6c

If the onLongPress start is removed as a callback on the PageListViewportGestures then the stutter is no longer noticeable.

**Theory:** The presence of the onLongTap recognizer in the Flutter gesture arena is not eliminated until the gesture exceeds a certain translation distance delta and/or enough time has passed. 

Is there a way to eliminate this delay with the recognizer, or make it as short as possible? We want the swiping gestures to be recognized as onScale as soon as possible.

**Comment by matthew-carroll** [comment]
I'm not aware of any way to both have that gesture recognizer, but not give it time to respond. It begs the question: what would a "long press" be, if you don't wait a "long" time?

To work around that timing issue, the first step is to define the desired behavior, which we might be able to implement with some kind of hack between the gesture arena and a non-competing `Listener`.

**Comment by dodgog** [comment]
The way flutter recognizes a long press gesture right now (correct me if I'm wrong) is:

> If preAcceptSlopTolerance of 18 pixels is not exceeded before 500ms pass, the long press recognizer wins.

The problem with this approach is the time lag associated with the time it takes a finger to move these 18 pixels. During that time the gesture is not recognized as panning.

The desired behavior minimizes that time lag. Right now, in the worst case a very-very slow movement at 36 pixels a second would be considered a long tap and would have an almost 500 ms lag. We either should set preAcceptSlopTolerance to a smaller value, I would try 5 pixels, or think of a more comprehensive way to rule out the long press gesture, so that it rules out based on speed and distance. I can think of this one, but it's a work in progress, and is probably too much:

> The onLongTap recognizer looses if at any point during the 500ms the speed of the finger gesture exceeds a set speed limit of 50 pixels/second or the continuous path length exceeds 10 pixels (alternatively the deviation from gesture start exceeds 10 pixels). This means conceptualizing slop not as a distance between the beginning of the gesture to the furthest deviating point, but rather as total path length for continuous parts of the gesture path. (Let me know if this explanation is comprehensible).

That way a very-very slow movement at 9.9 pixels a second would be considered a long tap and would take almost the entire 500 ms to recognize, but also a 99 pixels a second for 100ms would be caught.

I think that Apple's long tap logic seems to follow this intuition, because, of course, long tap cannot be preceded by panning; _fast_ and short swipes immediately get ruled out and we are not seeing the time lag.

**Comment by matthew-carroll** [comment]
> During that time the gesture is not recognized as panning.

This is correct. Because if the finger hasn't moved at least 18 pixels, we literally don't know whether its a pan or a long press. This brings me back to my original question: what would a "long press" be, if you don't wait a "long" time?

To me, this issue sounds like "We want immediate long press recognition, and we want immediate panning recognition, in the same space". But this desire doesn't make sense. If both gestures are possible, then you need to wait however long it takes to figure out which gesture is happening. Flutter can't read the user's mind, so Flutter can look at distance moved and time that's passed.

The post above includes a handful of seemingly arbitrary numbers. I have no idea if those would help or not. Have you tried applying those numbers? Is there some reason to think that your proposed numbers are better than Flutter numbers?

If you've found a better heuristic than Flutter, should this change be made in Flutter, itself, rather than this package?

**Comment by dodgog** [comment]
> To me, this issue sounds like "We want immediate long press recognition, and we want immediate panning recognition, in the same space".

I'm saying "desired behavior minimizes [the] time lag", I'm not saying we want immediate recognition. One gesture has victory in the arena. We want to brainstorm ways to make long presses loose faster because we have found that their recognizer in the arena creates a significant delay for slow scrolling gestures.

> The post above includes a handful of seemingly arbitrary numbers. 

I'm attempting to quantify the "faster". The numbers are not arbitrary, they are inspired by iOS delay times, which are never as long as 500ms, you can measure by counting high-speed video frames yourself. I believe it's possible to cut the worst case delay in the viewport at least in half without loosing any long presses.

I would like your feedback on how to make it faster. In my first comment I've brought up a potential way to quicker exclude the long presses from the arena. Question: Does it make sense? Can you think of a better decider?

Question: If one were to implement such decider, would it make sense to put it into the pageListViewportGestures methods, for example into onScale, so that onScale and onLongPress don't fight in the arena and the pageListViewport gets to decide which one persists OR is it better to override the gestureDetector and have the two gestures fight there?

You can also take it as an open-ended question and disregard my numbers, metrics, and approaches if you wish. We want smaller delay before slow page list scrolls.

**Comment by matthew-carroll** [comment]
> We want to brainstorm ways to make long presses loose faster because we have found that their recognizer in the arena creates a significant delay for slow scrolling gestures.

It's likely that your only robust option is to write a new gesture recognizer for long press, which more eagerly loses out to other gestures.

> Question: If one were to implement such decider, would it make sense to put it into the pageListViewportGestures methods, for example into onScale, so that onScale and onLongPress don't fight in the arena and the pageListViewport gets to decide which one persists OR is it better to override the gestureDetector and have the two gestures fight there?

I think you'll need to build a new recognizer to win out in the arena. It's not clear to me how one would effectively use the existing methods, called by the existing recognizers, to reduce the time.

**Comment by rutvik110** [comment]
I think its safe to say that what we want here is a custom long press recognizer, that will loose out early to other gestures with adjusted deadline period and other metrics and let other recognizers win early to reduce the lag(mentioned by @dodgog) seen in the shared demo.

@matthew-carroll Are there any things that you may've looked into earlier related to this that you can point out to help me get started on it. I'll also do some research on my side and see what I can find related to building custom recognizers such as the one that we want in this case and work out a proposed solution we can work from.

**Comment by matthew-carroll** [comment]
@rutvik110 the only info I have that might help is the custom gesture recognizer I wrote for Super Editor, which, itself, was based heavily on Flutter's existing multi-tap gesture recognizer. Even now it's unclear why some of the things in those recognizers need to exist. It's a very poorly documented area of Flutter. But if you compare the Super Editor recognizer to some of the built-in Flutter recognizers, you might be able to infer some of the responsibilities.

**Comment by matthew-carroll** [comment]
@rutvik110 any update on this? Can you prioritize this ticket?

**Comment by rutvik110** [comment]
Sry have been swamped with few other things so this got delayed. I'm taking look at it rn.
